### PR TITLE
feat: limited support of lamdas in `Series.apply`

### PR DESCRIPTION
**Pull request by shobsi**
BEGIN_COMMIT_OVERRIDE
feat: limited support of lambdas in `Series.apply` (#345)
END_COMMIT_OVERRIDE

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated https://screenshot.googleplex.com/6ZEiKXPz8LWMTRf

Partially fixes internal issue 295964341 🦕

### CODE REVIEW

File: third_party/bigframes_vendored/pandas/core/series.py [hunk 1]
```diff
@@ -1152,6 +1152,35 @@ def apply(
             >>> names = bpd.Series(["Alice", "Bob"])
             >>> hashes = names.apply(get_hash)
 
+        There is a limited support of simple functions and lambdas which can be
+        operated directly (without converting into a `remote_function`) on the
+        BigQuery DataFrames objects. This approach takes advantage of a nuance
+        in the way BigQuery DataFrames objects are modeled internally and works
+        only if the function body contains only arithmatic or logical operators.
```
Comment (tswast) [inline third_party/bigframes_vendored/pandas/core/series.py:1159]: I'd rather we rephrase this as "ufunc" emulation support, as defined in the pandas docs. https://pandas.pydata.org/docs/reference/api/pandas.Series.apply.html

File: bigframes/core/compile/scalar_op_compiler.py [hunk 2]
```diff
@@ -658,10 +658,6 @@ def isin_op_impl(x: ibis_types.Value, op: ops.IsInOp):
 
 @scalar_op_compiler.register_unary_op(ops.RemoteFunctionOp, pass_op=True)
 def remote_function_op_impl(x: ibis_types.Value, op: ops.RemoteFunctionOp):
-    if not hasattr(op.func, "bigframes_remote_function"):
```
Comment (tswast) [inline bigframes/core/compile/scalar_op_compiler.py:661]: This is the wrong layer at which to do this. Please instead call this on the actual Series object.

That way numpy ufuncs will also work as expected.

**Review by tswast — CHANGES_REQUESTED** [review]

File: bigframes/core/compile/scalar_op_compiler.py (hunk 2 above)
Comment (tswast) [reply bigframes/core/compile/scalar_op_compiler.py:661]: Also, please add tests for passing in numpy ufuncs directly in addition to the existing lambda tests.

File: third_party/bigframes_vendored/pandas/core/series.py (hunk 1 above)
Comment (shobsi) [reply third_party/bigframes_vendored/pandas/core/series.py]: Updated, PTAL.

File: bigframes/core/compile/scalar_op_compiler.py (hunk 2 above)
Comment (shobsi) [reply bigframes/core/compile/scalar_op_compiler.py]: Added tests with select numpy ufuncs supported by bigframes

File: bigframes/series.py [hunk 3]
```diff
@@ -1213,6 +1213,14 @@ def apply(self, func) -> Series:
         # Reproject as workaround to applying filter too late. This forces the filter
         # to be applied before passing data to remote function, protecting from bad
         # inputs causing errors.
+        if not callable(func):
+            raise ValueError(
+                "Only a ufunc (a NumPy function that applies to the entire Series) or a remote function that only works on single values are supported."
+            )
+
+        if not hasattr(func, "bigframes_remote_function"):
+            return func(self)
```
[the remark this answers is not in the document]
Comment (shobsi) [reply bigframes/series.py]: Done, thanks!

File: bigframes/series.py [hunk 4]
```diff
@@ -1216,6 +1222,22 @@ def apply(self, func) -> Series:
...
+
+        if not hasattr(func, "bigframes_remote_function"):
+            try:
+                return func(self)
+            except Exception as ex:
+                # This could happen if any of the operators in func is not
+                # supported on a Series. Let's guide the customer to use a
+                # remote function instead
+                if hasattr(ex, "message"):
+                    ex.message += "\n{_remote_function_recommendation_message}"
```
Comment (tswast) [inline bigframes/series.py:1238]: This needs to be an `f` string.

File: bigframes/series.py [hunk 5]
```diff
@@ -1216,6 +1222,22 @@ def apply(self, func) -> Series:
...
         # to be applied before passing data to remote function, protecting from bad
         # inputs causing errors.
+        if not callable(func):
+            raise ValueError(
+                "Only a ufunc (a NumPy function that applies to the entire Series) or a remote function that only works on single values are supported."
+            )
+
+        if not hasattr(func, "bigframes_remote_function"):
+            try:
+                return func(self)
```
Comment (tswast) [inline bigframes/series.py:1232]: Since it may result in incorrect values if this isn't a true vectorized function, let's check for `by_row=False`. If `by_row="compat"` (default) then raise and suggest either remote_function or by_row=False.

File: bigframes/series.py (hunk 4 above)
Comment (shobsi) [reply bigframes/series.py]: Thanks for catching, corrected.

File: bigframes/series.py (hunk 5 above)
Comment (shobsi) [reply bigframes/series.py:1245]: Done, PTAL.

File: bigframes/series.py [hunk 6]
```diff
@@ -1210,12 +1216,41 @@ def _groupby_values(
...
+        if not callable(func):
+            raise ValueError(
+                "Only a ufunc (a function that applies to the entire Series) or a remote function that only works on single values are supported."
+            )
+
+        if not hasattr(func, "bigframes_remote_function"):
+            # It is not a remote function
+            # Then it must be a vectorized function that applies to the Series
+            # as a whole
+            assert (
```
Comment (tswast) [inline bigframes/series.py:1240]: AssertionError is a bit of an odd one to raise. Usually that means some invariant that we never expect to happen has violated. Please raise `ValueError` instead.

Comment (shobsi) [reply bigframes/series.py]: Done, PTAL, thanks.

**Review by tswast — APPROVED** [review]
Love it, thanks!
